File: Gestor_Proyectos/src/main/java/com/GestorProyectos/entity/Developer.java
package com.GestorProyectos.entity;

import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Persisted developer profile.
 * Populated when a recruiter saves a candidate (Step 4).
 * Search results and profile fetches use DeveloperDto/DeveloperProfileDto (Redis cache).
 */
@Entity
@Table(name = "developer")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Developer {

    /** Composite key: "github:torvalds", "gitlab:user", etc. */
    @Id
    private String id;

    @Column(nullable = false)
    private String platform;

    @Column(nullable = false)
    private String username;

    private String displayName;
    private String avatarUrl;
    private String profileUrl;

    @Column(columnDefinition = "TEXT")
    private String bio;

    private String location;
    private String company;
    private String joinedAt;

    // ── Engagement ──────────────────────────────────────────────────────────
    private Integer followers;
    private Integer following;
    private Integer totalPublicRepos;
    private Integer ownedRepos;
    private Integer totalStars;
    private Integer totalForks;

    // ── Activity (from GitHub GraphQL, populated in Step 2) ─────────────────
    private Integer totalCommitsLastYear;
    @Column(name = "last90_days_activity")
    private Integer last90DaysActivity;

    @Column(name = "last30_days_activity")
    private Integer last30DaysActivity;
    private Integer totalPullRequests;
    private Integer totalIssues;
    private Integer totalPrReviews;

    // ── Languages (JSON: [{name, color, bytes, percentage}]) ────────────────
    @Column(columnDefinition = "TEXT")
    private String topLanguagesJson;

    // ── Scoring (computed by ScoringService, Step 3) ─────────────────────────
    private Integer score;          // weighted total 0-100
    private Integer scoreRecent;    // sub-score: recent activity
    private Integer scoreImpact;    // sub-score: project impact
    private Integer scoreContrib;   // sub-score: contributions
    private Integer scoreCollab;    // sub-score: collaboration
    private Integer scoreTech;      // sub-score: tech stack
    private String  scoreTier;      // Expert / Senior / Mid-level / Junior / Beginner

    private LocalDateTime fetchedAt;
}

File: frontend/src/lib/developer.ts
import { api } from "./api";

export interface LanguageStat {
  name: string;
  color: string;
  bytes: number;
  percentage: number;
}

export interface ContributionStats {
  totalCommits: number;
  totalPullRequests: number;
  totalIssues: number;
  totalPrReviews: number;
  totalContributions: number;
  last90DaysCommits: number;
  last30DaysCommits: number;
}

export interface ScoreBreakdown {
  recentActivity: number;
  projectImpact: number;
  contributions: number;
  collaboration: number;
  techStack: number;
  total: number;
  tier: string;
  aiSummary: string | null;
  aiInsights: string[] | null;
}

export interface DeveloperProfile {
  id: string;
  platform: string;
  username: string;
  displayName: string;
  avatarUrl: string;
  profileUrl: string;
  bio: string | null;
  location: string | null;
  company: string | null;
  joinedAt: string | null;
  email: string | null;
  blog: string | null;
  linkedinUrl: string | null;
  followers: number;
  following: number;
  totalPublicRepos: number;
  ownedRepos: number;
  totalStars: number;
  totalForks: number;
  topLanguages: LanguageStat[];
  contributionStats: ContributionStats;
  scoreBreakdown: ScoreBreakdown | null;
}

export async function getDeveloperProfile(
  platform: string,
  username: string
): Promise<DeveloperProfile> {
  const { data } = await api.get(
    `/api/developer/${platform}/${username}/profile`
  );
  return data.data as DeveloperProfile;
}
